forklift/pallet.py

```diff
@@ -0,0 +1,151 @@
+        self.data = 'udc_ofndr.dat'
+        self.dirty = None
+
+    def requires_processing(self):
+        if self.dirty is not None:
+            return self.dirty
+
+        #: if folder never exists, expect first run and the crate has changes
+        data = os.path.join(self.corrections, self.data)
+
+        if not os.path.exists(self.corrections) or not os.path.exists(data):
+            os.makedirs(self.corrections, exist_ok=True)
+            self.get_files()
+
+            self.data_stats = os.stat(data)
+
+            self.dirty = True
+
+            return True
+
+        #: get mtime from current file
+        self.data_stats = os.stat(data)
+        last_modified = self.data_stats.st_mtime
+
+        #: connect to sft, check mtime on remote file, replace if newer
+        self.dirty = self.get_files(mtime=last_modified)
+
+        return self.dirty
+
+    def process(self):
+        dat_file = os.path.join(self.corrections, self.data)
+        stats = os.stat(dat_file)
+        success = True
+
+        try:
+            self.log.info('converting offender data')
+            frame = pd.read_csv(
+                dat_file,
+                sep='|',
+                engine='python',
+                quoting=csv.QUOTE_NONE,
+                names=schema.FIELDS,
+                converters=schema.CONVERTERS,
+            ).iloc[:, :-1]
+
+            self.log.debug(frame.info())
+
+            truncate = sqlalchemy.text('TRUNCATE TABLE offenders')
+            add_shape = (
+                'UPDATE offenders SET '
+                '''shape = geography::STGeomFromText('POINT(' + CONVERT(varchar, offenders.x) + ' ' + CONVERT(varchar, offenders.y) + ')', 4326)'''
+            )
+
+            #: load new data
+            engine = sqlalchemy.create_engine(database.CONNECTION)
+
+            with engine.connect() as connection:
+                self.log.debug('trucating table')
+                connection.execution_options(autocommit=True).execute(truncate)
+
+                self.log.info('inserting offender')
+                frame.to_sql(
+                    'offenders',
+                    engine,
+                    if_exists='append',
+                    index=False,
+                    chunksize=5000,
+                )
+
+                self.log.debug('creating shape')
+                connection.execution_options(autocommit=True).execute(add_shape)
+        except Exception as e:
+            self.log.fatal(e)
+            self.success = (False, 'unable to read csv and write data to sql')
+            success = False
+        finally:
+            with open(dat_file, 'w') as dat:
+                dat.truncate()
+
+            if success:
+                #: reset access stats from truncate
+                os.utime(dat_file, (stats.st_atime, stats.st_mtime))
+            else:
+                #: reset stats to original because errors
+                os.utime(dat_file, (self.data_stats.st_atime, stats.st_mtime))
```

Comment:
This doesn't work if the pallet fails on its first run. Not sure how to handle that but maybe something to think about. Maybe check to see if the two modified times are the same and if so then delete the file?

Reply:
i think my update will fix this. i'll request a re review...